torchinductor/lowering.py

```diff
@@ -1598,11 +1598,15 @@ def mean(x, axis=None, keepdim=False, *, dtype=None):
         x = to_dtype(x, dtype)
     size = x.get_size()
     axis = _validate_reduction_axis(x, axis)
+    # compute in higher-precision until end of mean lowering
+    output_dtype = x.get_dtype()
+    if output_dtype in (torch.float16, torch.bfloat16):
+        x = to_dtype(x, torch.float)
```

Comment:
is this needed? Triton codegen automatically upcasts on read, so no low precision numbers should be present in the kernel

Reply:
I think you are right, this should happen automatically.

Reply:
Without this, you get a cast back to float16 before the division, and the test fails

Reply:
There should not be cast backs other than at the kernel boundaries, and division should happen in the same kernel.

Reply:
yea, a related issue is that a separate kernel is launched for just the divide.. not sure yet ATM why it is not being fused
```

from ctypes import c_void_p, c_long
import torch
from torch import empty_strided, as_strided
from torchinductor.codecache import CppCodeCache, TritonCodeCache

aten = torch.ops.aten

import triton
import triton.language as tl

from torchinductor.triton_ops.autotune import pointwise_heuristics
from torchinductor.triton_ops.autotune import reduction_heuristics
from torchinductor.triton_ops.autotune import grid

@reduction_heuristics(size_hints=[128, 32768])
@triton.jit
def kernel0(in_ptr0, out_ptr0, xnumel, rnumel, XBLOCK : tl.constexpr, RBLOCK : tl.constexpr):
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.reshape(tl.arange(0, XBLOCK), [XBLOCK, 1])
    xmask = xindex < xnumel
    rbase = tl.reshape(tl.arange(0, RBLOCK), [1, RBLOCK])
    x0 = xindex
    _tmp1 = tl.zeros([XBLOCK, RBLOCK], tl.float32) + 0
    for roffset in range(0, rnumel, RBLOCK):
        rindex = roffset + rbase
        rmask = rindex < rnumel
        r1 = rindex
        tmp0 = tl.load(in_ptr0 + (r1 + (29040*x0)) % 3717120, xmask & rmask).to(tl.float32)
        _tmp1 = tl.where(xmask & rmask, _tmp1 + tmp0, _tmp1)
    tmp1 = tl.reshape(tl.sum(_tmp1, 1), [XBLOCK, 1])
    tl.store(out_ptr0 + x0, tmp1, xmask)

@reduction_heuristics(size_hints=[1, 128])
@triton.jit
def kernel1(in_ptr0, out_ptr0, xnumel, rnumel, XBLOCK : tl.constexpr, RBLOCK : tl.constexpr):
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.reshape(tl.arange(0, XBLOCK), [XBLOCK, 1])
    xmask = xindex < xnumel
    rbase = tl.reshape(tl.arange(0, RBLOCK), [1, RBLOCK])
    _tmp1 = tl.zeros([XBLOCK, RBLOCK], tl.float32) + 0
    for roffset in range(0, rnumel, RBLOCK):
        rindex = roffset + rbase
        rmask = rindex < rnumel
        r0 = rindex
        tmp0 = tl.load(in_ptr0 + r0, rmask)
        _tmp1 = tl.where(xmask & rmask, _tmp1 + tmp0, _tmp1)
    tmp1 = tl.reshape(tl.sum(_tmp1, 1), [XBLOCK, 1])
    tl.store(out_ptr0 + 0 + tl.zeros(tmp1.shape, tl.int32), tmp1, xmask)

@pointwise_heuristics(size_hints=[1])
@triton.jit
def kernel2(in_ptr0, out_ptr0, xnumel, XBLOCK : tl.constexpr):
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.reshape(tl.arange(0, XBLOCK), [XBLOCK])
    xmask = xindex < xnumel
    tmp0 = tl.load(in_ptr0 + 0, None).to(tl.float32)
    tmp1 = 3717120
    tmp2 = tmp0 / tmp1
    tl.store(out_ptr0 + 0 + tl.zeros(tmp2.shape, tl.int32), tmp2, xmask)


def call(arg0):
    arg0_size = arg0.size()
    s0 = arg0_size[0]
    s1 = arg0_size[1]
    s2 = arg0_size[2]
    buf0 = empty_strided((128, ), (1, ), device='cuda', dtype=torch.float32)
    kernel0[grid(128)](arg0, buf0, 128, 29040)
    buf1 = empty_strided((), (), device='cuda', dtype=torch.float16)
    kernel1[grid(1)](buf0, buf1, 1, 128)
    del buf0
    buf2 = empty_strided((), (), device='cuda', dtype=torch.float16)
    kernel2[grid(1)](buf1, buf2, 1)
    return (buf2, )


if __name__ == "__main__":
    from torchdynamo.testing import rand_strided
    from microbenchmarks.microbench import print_performance
    arg0 = rand_strided((10, 3, 352, 352), (371712, 123904, 352, 1), device='cuda', dtype=torch.float16)
    print_performance(lambda: call(arg0))
```

Reply:
Hm that's a problem if that's the case, I hoped #446 would help with that but probably it requires setting the number of tiles to work. Then all reduction lowerings, not just mean, will need to be fixed for this. 
Primtorch decomps handle casting to/from higher datatypes, so alternatively we can rely on primtorch traces inserting the necessary casts.

Reply:
Well, primtorch has the same issue of converting back to fp16 before the division

Reply:
For `mean` primtorch keeps reduction result in promoted type, and converts only after division https://github.com/pytorch/pytorch/blob/d6cd130e0b791a0556ebc7bbd6871840b65ad35c/torch/_refs/__init__.py#L1753